R/req-perform-multi.R

```diff
@@ -0,0 +1,228 @@
+#' Perform a multi request
+#'
+#' Perform a request with a multi policy, e.g. created by [req_paginate()] or
+#' [req_chunk()].
+#'
+#' @inheritParams req_perform
+#' @param max_requests The maximum number of requests to perform.
+#' @param cancel_on_error Should all pending requests be cancelled when you
+#'   hit an error. Set this to `TRUE` to stop all requests as soon as you
+#'   hit an error. Responses that were never performed will have class
+#'   `httr2_cancelled` in the result.
+#' @param progress Display a progress bar? Use `TRUE` to turn on a basic progress
+#'   bar, use a string to give it a name, or see [progress_bars] for more details.
+#'
+#' @return The result of `vec_c()`ing together the `data` fields that were
+#'   extracted by the `parse_resp` argument of [req_paginate()].
+#'   If `parse_resp` is not specified, it will be a list of the raw responses.
+#' @export
+#'
+#' @examples
+#' page_size <- 150
+#'
+#' req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_url_query(limit = page_size) %>%
+#'   req_paginate_next_url(
+#'     parse_resp = function(resp) {
+#'       parsed <- resp_body_json(resp)
+#'       results <- parsed$results
+#'       data <- data.frame(
+#'         name = sapply(results, `[[`, "name"),
+#'         url = sapply(results, `[[`, "url")
+#'       )
+#'
+#'       list(data = data, next_url = parsed$`next`)
+#'     },
+#'     n_pages = function(parsed) {
+#'       total <- parsed$count
+#'       ceiling(total / page_size)
+#'     }
+#'   )
+#'
+#' responses <- req_perform_multi(req_pokemon)
+req_perform_multi <- function(req,
+                              # paths = NULL,
+                              max_requests = NULL,
+                              cancel_on_error = FALSE,
+                              progress = TRUE,
+                              error_call = current_env()) {
+  # TODO support `path` argument
+  # * either a list of requests (same size as `n_requests`)
+  # * a path pattern
+
+  check_request(req)
+  check_has_multi_policy(req)
+  check_number_whole(max_requests, allow_infinite = TRUE, allow_null = TRUE, min = 1)
+  max_requests <- max_requests %||% Inf
+
+  n_requests <- min(req$policies$multi$n_requests, max_requests)
+  get_n_requests <- req$policies$multi$get_n_requests
+
+  # the implementation below doesn't really support an infinite amount of pages
+  # but 100e3 should be plenty
+  if (is.infinite(n_requests)) {
+    n_requests <- 100e3
+  }
+
+  perform <- error_wrapper(
+    f = req_perform,
+    cancel_on_error = cancel_on_error,
+    error_message = "When performing request {i}.",
+    error_class = "httr2_failure",
+    error_call = error_call
+  )
+
+  parse_resp <- error_wrapper(
+    f = function(resp) {
+      parse_resp <- req$policies$parse_resp
+      parse_resp(resp)
+    },
+    cancel_on_error = cancel_on_error,
+    error_message = "When parsing response {i}.",
+    error_class = "httr2_parse_failure",
```

Comment:
Or would you prefer to have the same class when the request fails and when the parsing fails? I.e. <httr2_failure> in both cases.